### PR TITLE
Prototype: Multiple language support with Rust

### PR DESCRIPTION
**Pull request by andygrunwald**
## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [X] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [X] I have read the **CONTRIBUTING** document.
* [X] `make ci` passes on my machine.
* [X] My change requires a change to the documentation.
* [X] I have updated the documentation accordingly.
* [ ] I have added tests to cover my changes.

## Why is it useful?

This is a prototype version of #496.
This version is far from complete, but it should start a first discussion or feedback round.
The feedback round is important to let me / us know if it is useful to put more effort into it or to drop the idea at all.

This PR shows how this can be done for Rust.
Testing would be welcomed.

## What it does?

It compiles a rust app with `cargo build` and adds it to the Artifacts.

## How can you test it?

1. Install Rust (e.g. via https://rustup.rs/)
2. Kickstart a new project via `cargo new --bin hello`
3. Place a new `.goreleaser.yaml` with this content into the project
```
rust:
- target:
  - x86_64-apple-darwin
  binary: hello

archive:
  format: tar.gz
  name_template: "{{ .ProjectName }}_{{ .Version }}_{{ .Os }}"
```
4. Tag it and release it like every other tool

## What is missing?

* [ ] Unit tests
* [ ] Proper testing with other pipes

## Challenges

* [ ] Goos, Goarch and Goarm are spreaded accross the codebase and used in a specific way (e.g. for templating). In the Rust context this adds a challenge, because targets are not splitted into the specific parts. They are represented like one string (e.g. `x86_64-apple-darwin` for Mac) known from LLVM
* [ ] Copy / Paste code: Right now we just copied many of the code parts. Reusage would make sense here

### CODE REVIEW

**Comment by caarlos0** [comment]
I like the idea but I'm not sure we should do it this way.

goreleaser already does too much (as someone else already stated in twitter).

I'm thinking that maybe we should have some sort pluggable pipes...

examples:

- a `piper` that build for rust (or maybe in this case it could even be a `builder` instead of a `piper`
- a `piper` that publishes new releases to twitter
- a `piper` that generates the `| sh` shell scripts

I don't yet know how to make all this though.

I like the way terraform works, for example, and I think it's a good place to start studying the implementation...

**Comment by caarlos0** [comment]
what do you think?

**Comment by caarlos0** [comment]
thinking about a `builder` interface, we could have an interface like this:

```go
type Builder interface {
	Build(ctx *context.Context, build config.Build, target buildtarget.Target) error
}
```

and then on the build pipe, something like:

```go
var builder Builder
switch lang {
case "go":
	builder = gobuilder.Default
case "rust":
	builder = rustbuilder.Default
}
```

or maybe even better: builders register themselves:

```go
package gobuilder

func init() {
	build.Register("go", Default)
}
```

and we n do something like this in the build pipe:

```go
build.ForLang(lang).Build(/* ... */)
```

what do you think?

**Comment by caarlos0** [comment]
I've prototyped something on #522 

check it out @andygrunwald !

**Comment by caarlos0** [comment]
about the goos, goarch and goarm on the artifact struct...

I think we can change them to uname-relatives. They are used to filter artifacts on other pipes (eg: docker, fpm, brew).

question is: can we break the rust `target` into os + arch + arm?

**Comment by caarlos0** [comment]
> question is: can we break the rust target into os + arch + arm reliably?

looking into the results of `rustc --print target-list`, seems like we can!

File: pipeline/rust/rust.go [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+
+func buildWithDefaults(ctx *context.Context, rust config.Rust) config.Rust {
+	if rust.Binary == "" {
+		rust.Binary = ctx.Config.Release.GitHub.Name
+	}
+
+	// One idea would be, if no rust.Target are defined
+	// to add the default target for the current OS.
+	// E.g. hit `rustup show`. The default host is shown.
+	// For an up to date Mac is will show "Default host: x86_64-apple-darwin"
```
Comment (caarlos0) [inline pipeline/rust/rust.go]: I think we could default this to the same targets as go:

- x86_64-apple-darwin
- x86_64-unknown-linux-gnu
- i686-apple-darwin
- i686-unknown-linux-gnu

**Comment by andygrunwald** [comment]
> goreleaser already does too much (as someone else already stated in twitter).

I agree here. When we continue this road, it will become hard to maintain.

I like your idea of the "builder" and "piper" examples.
To generalize it even more: What about languages that are target independent? I think about JavaScript, Python (e.g. egg files), PHP (composer, phar), Java, Kotlin, Scala (JVM).
They have only one target.

> I like the way terraform works, for example, and I think it's a good place to start studying the implementation...

Never looked deep into it. So at the moment, i am not a help here :(

> about the goos, goarch and goarm on the artifact struct...
> question is: can we break the rust target into os + arch + arm reliably?
> looking into the results of rustc --print target-list, seems like we can!

Those targets are not targets created by rust. Those are targets from LLVM.
I didn't found a document what describes how to parse this, but there seems to be some kind of rule. A few examples:

aarch64-linux-android
aarch64-unknown-freebsd
aarch64-unknown-fuchsia
aarch64-unknown-linux-gnu
aarch64-unknown-linux-musl
asmjs-unknown-emscripten
i586-unknown-linux-gnu

Here are some slides that named it "target-triple" with `Target Triple: <arch><sub>-<vendor>-<sys>-<abi>`: https://llvm.org/devmtg/2014-04/PDFs/LightningTalks/2014-3-31_ClangTargetSupport_LighteningTalk.pdf

Maybe @mre or @timglabisch knows more?

File: pipeline/rust/rust.go (hunk 1 above)
Comment (andygrunwald) [reply pipeline/rust/rust.go]: We thought about this as well. Like If MacOSx than target this. If linux, then this and so on. But we have to take the architecture and even the linker-system into account (when it comes to GNU libc or Musl libc). And there it is getting hard to decide what is the default

**Comment by mre** [comment]
Yup, target triples. More on that [here](https://github.com/rust-lang-nursery/rust-forge/blob/master/platform-support.md) and [here](https://github.com/japaric/rust-cross). 

> Triples are usually formatted as follows: {arch}-{vendor}-{sys}-{abi}. 
> For example, the triple arm-unknown-linux-gnueabihf refers to the systems that share these characteristics:
>  *  architecture: arm.
>  * vendor: unknown. In this case, no vendor was specified and/or is not important.
>  * system: linux.
>  * ABI: gnueabihf. gnueabihf indicates that the system uses glibc as its C standard library (libc) implementation and has hardware accelerated floating point arithmetic (i.e. an FPU).

The attentive reader might have noticed, that triples are not necessarily *triples*. 

Also - contrary to Go - Rust can not cross compile to all platforms on every platfrom. For example, it's [quite tricky to compile to Mac from a Linux machine](https://github.com/japaric/rust-cross#i-want-to-build-binaries-for-linux-mac-and-windows-how-do-i-cross-compile-from-linux-to-mac). Most people use external services like Travis or AppVeyor for that. There's also [trust](https://github.com/japaric/trust), which provides templates for these services to make cross compiling easier, but it still requires manual steps.

So, I guess if we want to support Rust, we should be a bit more conservative with the build targets and only use the host triple as a default. We can always add more targets later if possible.

File: pipeline/rust/rust.go (hunk 1 above)
Comment (mre) [reply pipeline/rust/rust.go]: Won't be easy. See comment [here](https://github.com/goreleaser/goreleaser/pull/520#issuecomment-359570387).

**Comment by caarlos0** [comment]
so I guess we can default to the current "triplet" only...

**Comment by andygrunwald** [comment]
@mre @caarlos0 How do we continue here?
#522 is merged (🎉 ), this was the basement for this, right?

Should we (i, @mre, @timglabisch) finish this (with your help)?
Would you (@caarlos0) build the first implementation to show how it is done?

How did we align with the triplet defaults? Or with the native `cargo` cmd?
@mre You mentioned something related to "Build Rust inside a Docker container" for Linux, right? Can you elaborate on this or show an example?
And do you see this as a must-have? Or do you consider this as a user problem?

**Comment by mre** [comment]
> @mre You mentioned something related to "Build Rust inside a Docker container" for Linux, right? Can you elaborate on this or show an example?

There's [cross](https://github.com/japaric/cross) and [cargo-docker](https://github.com/DenisKolodin/cargo-docker), which both use Docker for cross-compiling Rust code.

> And do you see this as a must-have? Or do you consider this as a user problem?

It's completely optional, I guess. At a future stage, we could piggy-back on the above tools for cross-compling, if they are installed. Alternatively, we could integrate a Docker pipeline to be completely independent. A sort of middle-ground would be having our own public (Rust) images that can be used for cross-compling.

**Comment by andygrunwald** [comment]
> Alternatively, we could integrate a Docker pipeline to be completely independent. A sort of middle-ground would be having our own public (Rust) images that can be used for cross-compling.

You mean a `docker-rust`? or a `docker` lang impl? this would be smart.

**Comment by mre** [comment]
> You mean a docker-rust? or a docker lang impl? this would be smart.

Was thinking about a pure, language-agnostic `docker` pipeline. The user would define, which image will be used for compiling. Mostly, people might want to use the official images like [rust](https://hub.docker.com/_/rust/) or [go](https://hub.docker.com/_/golang/) for cross-compiling. Yes, even for Go it makes sense because not everybody might want to install the golang toolchain on their build server. Same for PHP + composer or Javascript + Yarn.
Just an idea. 😜

**Comment by caarlos0** [comment]
> #522 is merged (🎉 ), this was the basement for this, right?

yeap =D



> Should we (i, @mre, @timglabisch) finish this (with your help)?
> Would you (@caarlos0) build the first implementation to show how it is done?

I think you we can kind of do it together: change this to be a builder impl, we can change what else needs to be changed while doing it.

Meanwhile, I'll experiment with plugins and etc

**Comment by caarlos0** [comment]
> Was thinking about a pure, language-agnostic docker pipeline. The user would define, which image will be used for compiling. Mostly, people might want to use the official images like rust or go for cross-compiling. Yes, even for Go it makes sense because not everybody might want to install the golang toolchain on their build server. Same for PHP + composer or Javascript + Yarn.
> Just an idea. 😜

like this idea as well =D

we could have a `rust` lang and a more generic `docker` lang (or `docker-go`, `docker-rust` etc)

**Comment by mre** [comment]
> we could have a rust lang and a more generic docker lang (or docker-go, docker-rust etc)

Exactly. A pre-configured rust toolchain using cargo that covers most use-cases and a language-agnostic Docker toolchain for total control, which can be used for cross-compiling.

**Comment by caarlos0** [comment]
or maybe we can do the plugin thing later (hopefully go will have native plugins for all platforms soon enough?)?

**Comment by guidiego** [comment]
Some news about use goreleaser + cargo?

**Comment by caarlos0** [comment]
> @caarlos0 @mre What do you think?

no problem on my side

**Comment by caarlos0** [comment]
> Some news about use goreleaser + cargo?

not at the moment :/

**Comment by andygrunwald** [comment]
From my side as well no progress so far.
Time is quite limited at the moment, to less time to wrap my head around :(
Sorry.

**Comment by caarlos0** [comment]
> From my side as well no progress so far.
> Time is quite limited at the moment, to less time to wrap my head around :(
> Sorry.

no problem, take your time :)

**Comment by caarlos0** [comment]
will close this for now.

I'll try to work on a plugin system in the next few days.

Thanks!

**Pull request closed without merging**